src/tensor_ops/upscale2d/upscale2d.cu

```diff
@@ -0,0 +1,243 @@
+#include "cuda_utils.cuh"
+
+struct Upscale2dOp {
+    size_t batch;
+    size_t chan;
+    size_t h_in;
+    size_t h_out;
+    size_t w_in;
+    size_t w_out;
+};
+
+template<typename T>
+__device__ void nearest_upscale2d_fwd(
+    const Upscale2dOp op,
+    const size_t *inp_strides,
+    const size_t *out_strides,
+    const T *inp, // 4d (Batch, Channels, Height, Width)
+    T *out // 4d (Batch, Channels, HeightOut, WidthOut)
+) {
+    unsigned int i = blockIdx.x * blockDim.x + threadIdx.x;
+    const size_t numel = op.batch * op.chan * op.h_out * op.w_out;
+    if (i >= numel) {
+        return;
+    }
+
+    float h_scale = ((float)(inp_strides[1]/inp_strides[2]))/(out_strides[1]/out_strides[2]);
+    float w_scale = ((float)(inp_strides[2]/inp_strides[3]))/(out_strides[2]/out_strides[3]);
+
+    unsigned int idx = i;
+    const size_t ow = idx % op.w_out;
+    idx /= op.w_out;
+    const size_t oh = idx % op.h_out;
+    idx /= op.h_out;
+    const size_t c = idx % op.chan;
+    idx /= op.chan;
+    const size_t b = idx % op.batch;
+    idx /= op.batch;
+
+    size_t ih = h_scale * oh;
+    size_t iw = w_scale * ow;
+
+    size_t inp_i = b * inp_strides[0] + c * inp_strides[1] + ih * inp_strides[2] + iw * inp_strides[3];
+    
+    out[i] = inp[inp_i];
+}
+
+template<typename T>
+__device__ void nearest_upscale2d_bwd(
+    const Upscale2dOp op,
+    const size_t *inp_strides,
+    const size_t *out_strides,
+    const T *inp, // 4d (Batch, Channels, Height, Width)
+    T *grad_inp,
+    const T *out, // 4d (Batch, Channels, HeightOut, WidthOut)
+    const T *grad_out
+) {
+    unsigned int i = blockIdx.x * blockDim.x + threadIdx.x;
+    const size_t numel = op.batch * op.chan * op.h_in * op.w_in;
+    if (i >= numel) {
+        return;
+    }
+
+    float h_scale = ((float)(inp_strides[1]/inp_strides[2]))/(out_strides[1]/out_strides[2]);
+    float w_scale = ((float)(inp_strides[2]/inp_strides[3]))/(out_strides[2]/out_strides[3]);
+
+    unsigned int idx = i;
+    const size_t x = idx % op.w_in;
+    idx /= op.w_in;
+    const size_t y = idx % op.h_in;
+    idx /= op.h_in;
+    const size_t c = idx % op.chan;
+    idx /= op.chan;
+    const size_t b = idx % op.batch;
+    idx /= op.batch;
+
+    // Probably isn't efficient, but it works
+    size_t oh_s = 0;
+    size_t ow_s = 0;
+    size_t oh_e = op.h_out-1;
+    size_t ow_e = op.w_out-1;
+    while (oh_s*h_scale < y) {
+        oh_s++;
+    }
```

Comment:
Is this `size_t oh_s = y / h_scale + 1`?

Reply:
I dont think so if h_scale divides y.